src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,322 @@ namespace System.Net.Sockets
+                    Interop.Sys.Close((IntPtr)_shutdownWritePipe);
+                }
+                if (_buffer != null)
+                {
+                    Interop.Sys.FreeSocketEventBuffer(_buffer);
+                }
+                if (_port != -1)
+                {
+                    Interop.Sys.CloseSocketEventPort(_port);
+                }
+                throw;
+            }
+        }
+
+        private void EventLoop()
+        {
+            try
+            {
+                bool shutdown = false;
+                while (!shutdown)
+                {
+                    int numEvents = EventBufferCount;
+                    Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
+                    if (err != Interop.Error.SUCCESS)
                     {
-                        context.HandleEvents(_buffer[i].Events);
+                        throw new InternalException();
+                    }
+
+                    // The native shim is responsible for ensuring this condition.
+                    Debug.Assert(numEvents > 0);
+
+                    for (int i = 0; i < numEvents; i++)
+                    {
+                        IntPtr handle = _buffer[i].Data;
+                        if (handle == ShutdownHandle)
+                        {
+                            shutdown = true;
+                        }
+                        else
+                        {
+                            SocketAsyncContext context = GetContextFromHandle(handle);
+                            if (context != null)
+                            {
+                                context.HandleEvents(_buffer[i].Events);
+                            }
+                        }
                     }
                 }
+
+                EndShutdown();
+            }
+            catch (Exception e)
+            {
+                Environment.FailFast("Exception thrown from SocketAsyncEngine event loop", e);
+            }
+        }
+
+        private void BeginShutdown()
+        {
+            //
+            // Write to the pipe, which will wake up the event loop and cause it to exit.
+            //
+            byte* buf = stackalloc byte[1];
+            int bytesWritten = Interop.Sys.Write(_shutdownWritePipe, buf, 1);
+            if (bytesWritten != 1)
+            {
+                throw new InternalException();
             }
         }
 
-        public bool TryRegister(SafeCloseSocket socket, Interop.Sys.SocketEvents current, Interop.Sys.SocketEvents events, GCHandle handle, out Interop.Error error)
+        private void EndShutdown()
+        {
+            //
+            // The event loop is exiting, so we can free all native resources.
+            //
+            Interop.Sys.Close((IntPtr)_shutdownReadPipe);
+            Interop.Sys.Close((IntPtr)_shutdownWritePipe);
+            Interop.Sys.CloseSocketEventPort(_port);
+            Interop.Sys.FreeSocketEventBuffer(_buffer);
```

Comment:
This is almost the same as the catch block when starting the event loop.  Could those checks be added here, and then this function be reused there?